Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -17,296 +17,526 @@
+    ComPtr<ID2D1DrawingStateBlock> d2dState;
+
+    // Fills
+    ComPtr<ID2D1Brush> fillBrush;
+
+    // Strokes
+    ComPtr<ID2D1Brush> strokeBrush;
+    ComPtr<ID2D1StrokeStyle> strokeStyle;
+    D2D1_STROKE_STYLE_PROPERTIES strokeProperties{
+        D2D1_CAP_STYLE_FLAT,
+        D2D1_CAP_STYLE_FLAT,
+        D2D1_CAP_STYLE_FLAT,
+        D2D1_LINE_JOIN_MITER,
+        10.f, // Default from Reference Docs
+        D2D1_DASH_STYLE_SOLID,
+        0.f,
+    };
+    std::vector<CGFloat> dashes;
+
+    CGFloat lineWidth = 1.0f;
+
+    CGFloat flatness;
+
+    // Clipping
+    CGMutablePathRef currentClippingPath;
+
+    // Image Drawing
+    D2D1_INTERPOLATION_MODE bitmapInterpolationMode;
+
+    // Userspace Coordinate Transformation
+    CGAffineTransform transform = CGAffineTransformIdentity;
+    CGAffineTransform textMatrix = CGAffineTransformIdentity;
+
+    // Alpha Blending
+    CGFloat alpha = 1.0f;
+};
+
+struct __CGContext {
+    struct __CGContextImpl {
+        ComPtr<ID2D1RenderTarget> renderTarget;
+
+        // Calculated at creation time, this transform flips CG's drawing commands,
+        // anchored in the bottom left, to D2D's top-left coordinate system.
+        CGAffineTransform cgCompatibilityTransform;
+
+        std::stack<__CGContextDrawingState> stateStack;
+
+        woc::unique_cf<CGMutablePathRef> currentPath;
+
+        bool allowsAntialiasing;
+        bool allowsFontSmoothing;
+        bool allowsFontSubpixelPositioning;
+        bool allowsFontSubpixelQuantization;
+
+        __CGContextImpl() : stateStack() {
+            // Set up a default/baseline state
+            stateStack.emplace();
+        }
+    };
+
+    CFRuntimeBase _base;
+
+    // Use an inline impl struct so that we don't need to placement new each element.
+    __CGContextImpl _impl;
+
+    inline ComPtr<ID2D1RenderTarget>& RenderTarget() {
+        return _impl.renderTarget;
+    }
 
-#define DEBUG_CONTEXT_COUNT
+    inline std::stack<__CGContextDrawingState>& GStateStack() {
+        return _impl.stateStack;
+    }
 
-int contextCount = 0;
-static IWLazyClassLookup _LazyUIImage("UIImage"), _LazyUIScreen("UIScreen");
+    inline __CGContextDrawingState& CurrentGState() {
+        return GStateStack().top();
+    }
 
-pthread_mutex_t _cairoLock = PTHREAD_MUTEX_INITIALIZER;
+    inline void ClearPath() {
+        _impl.currentPath.reset();
+    }
 
-@interface CGNSContext : _CGLifetimeBridgingType
-@end
+    inline ComPtr<ID2D1Factory> Factory() {
+        ComPtr<ID2D1Factory> factory;
+        _impl.renderTarget->GetFactory(&factory);
+        return factory;
+    }
+};
 
-@implementation CGNSContext
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wobjc-missing-super-calls"
-- (void)dealloc {
-    delete (__CGContext*)self;
+#pragma region CFRuntimeClass
+static Boolean __CGContextEqual(CFTypeRef cf1, CFTypeRef cf2) {
+    return FALSE;
```

Comment:
Nope! Looks like CG defaults to these two.

---
In reply to: [82916541](https://github.com/Microsoft/WinObjC/pull/1126#discussion_r82916541) [](ancestors = 82916541)